python/ray/util/client/logsclient.py

```diff
@@ -44,21 +45,27 @@ def _log_main(self) -> None:
                     self.stdstream(level=record.level, msg=record.msg)
                 self.log(level=record.level, msg=record.msg)
         except grpc.RpcError as e:
-            if e.code() == grpc.StatusCode.CANCELLED:
-                # Graceful shutdown. We've cancelled our own connection.
-                logger.info("Cancelling logs channel")
-            elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                              grpc.StatusCode.RESOURCE_EXHAUSTED):
-                # TODO(barakmich): The server may have
-                # dropped. In theory, we can retry, as per
-                # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-                # in practice we may need to think about the correct semantics
-                # here.
-                logger.info("Server disconnected from logs channel")
-            else:
-                # Some other, unhandled, gRPC error
-                logger.exception(
-                    f"Got Error from logger channel -- shutting down: {e}")
+            self._process_rpc_error(e)
+
+    def _process_rpc_error(self, e: grpc.RpcError):
```

Comment:
Would it make sense to put the common code between logsclient and dataclient here into common.py too?

Reply:
+1 I hope we can create an abstraction for RPC stream processing in the client, if possible. The logic between dataclient and logsclient diverge a bit after reconnection, but maybe the high level structure is similar enough?

Reply:
The main PR will have separate error processing in the reconnect PR (the dataclient is responsible for resetting the channel if it can't get pings through, while the logsclient will wait for the dataclient to do so), so I think it's fine to leave them separate for now.